Add DeepSeek-v3 workload

parthpower opened the pull request:
# What's new?

Adds `MockedModel` for DeepSeek-v3 based on open source model's code[^1], DeepGEMM[^4] and technical report paper[^2]

# Details

## `MockedDeepSeek`

mostly based on `MockedMegatron` for DP, TP comms.

- add `DeepSeekV3Model` top-level module
- add `DeepSeekMoE` for MoE with shared-experts. Comms to use FP8 dispatch and BF16 combine.[^2]
- add `DeepSeekMLA` for Multi-Head Latent Attention (MLA) layer
  - use compression matrix[^2], it's mentioned as LoRA in DeepSeek-v3's `model.py`[^3][^5]
  - decoupled rotary positional embedding (i.e. RoPE)

## `AiobDeepSeek`

modifies `AiobMegatron` to include FP8 matmul kernels from DeepGEMM[^4] to get compute times for experts

**Update 9/12/2025**:  Add `DeepSeekMLA` support in AIOB https://github.com/aliyun/aicb/pull/41/commits/588c201d2c44caf466a94c62398fe2ed867e5929

# Generated Workload

## GPU Workload with AIOB

```bash
sh scripts/megatron_gpt.sh \
  --frame DeepSeek \
  --tensor_model_parallel_size 4 \
  --pipeline_model_parallel 1 \
  --moe_enable \
  --expert_model_parallel_size 1 \
  --global_batch 4 \
  -m deepseek \
  --num_experts 4 \
  --micro_batch 1 \
  --sp --swiglu --world_size 4 --num_layers 10 --aiob_enable
```

Workload CSV file: [megatron_DeepSeek_sp_True_iteration_1_computationEnable_True_4n_log.csv](https://github.com/user-attachments/files/21850121/megatron_DeepSeek_sp_True_iteration_1_computationEnable_True_4n_log.csv)

## SimAI workload with AIOB

In the AIOB timing, note that `mlp_*` times for DeepSeek's MoE is much higher compared to that of Megatron even with FP8, that's mostly because of the extra linear layer in DeepSeek's Expert

i.e.

```python
# DeepSeek's Expert
self.w2(F.silu(self.w1(x)) * self.w3(x))

# vs Megatron's Expert
self.w2(act_fn(self.w1(x))
```

```bash
sh scripts/megatron_workload_with_aiob.sh \
-m deepseek16 --world_size 2048 --tensor_model_parallel_size 1 --pipeline_model_parallel 1 --sp --ep 16 \
--moe_router_topk 6 --moe_enable  \
--frame DeepSeek --global_batch 4096 \
--micro_batch 1 --seq_length 4096 --swiglu \
--aiob_enable
```

SimAI workload: 
[None-DeepSeek_16B-world_size2048-tp1-pp1-ep16-gbs4096-mbs1-seq4096-MOE-True-GEMM-False-flash_attn-False.txt](https://github.com/user-attachments/files/21850156/None-DeepSeek_16B-world_size2048-tp1-pp1-ep16-gbs4096-mbs1-seq4096-MOE-True-GEMM-False-flash_attn-False.txt)

DeepSeek16 AIOB Compute times on H100: 
[DeepSeek_16B-world_size2048-tp1-pp1-ep16-gbs4096-mbs1-seq4096-flash_attn-False.txt](https://github.com/user-attachments/files/21850178/DeepSeek_16B-world_size2048-tp1-pp1-ep16-gbs4096-mbs1-seq4096-flash_attn-False.txt)



# Citations

```
@misc{deepseekai2024deepseekv3technicalreport,
      title={DeepSeek-V3 Technical Report}, 
      author={DeepSeek-AI},
      year={2024},
      eprint={2412.19437},
      archivePrefix={arXiv},
      primaryClass={cs.CL},
      url={https://arxiv.org/abs/2412.19437}, 
}
```

[^1]: https://github.com/deepseek-ai/DeepSeek-V3/tree/f6e34dd26772dd4a216be94a8899276c5dca9e43

[^2]: [DeepSeek-V3 Technical Report](https://arxiv.org/abs/2412.19437)

[^3]: https://github.com/deepseek-ai/DeepSeek-V3/blob/f6e34dd26772dd4a216be94a8899276c5dca9e43/inference/model.py#L42-L43

[^4]: https://github.com/deepseek-ai/DeepGEMM/tree/391755ada0ffefa9a6a52b6f14dcaf22d1a463e0

[^5]: ![deepseek-aicb-mla drawio](https://github.com/user-attachments/assets/149e0d35-386b-4d10-8b8a-537abe030e00)

parthpower committed:
feat(deepseek): add mocked model for DeepSeek-v3

add mocked_model.MockedDeepSeek based off from
mocked_model.MockedMegatron

MOE: include shared experts

MLA: Multi-Head Latent Attention with down-projection matrix and
decoupled query RoPE

based on
https://github.com/deepseek-ai/DeepSeek-V3/tree/f6e34dd26772dd4a216be94a8899276c5dca9e43

and the paper:

@misc{deepseekai2024deepseekv3technicalreport,
      title={DeepSeek-V3 Technical Report},
      author={DeepSeek-AI},
      year={2024},
      eprint={2412.19437},
      archivePrefix={arXiv},
      primaryClass={cs.CL},
      url={https://arxiv.org/abs/2412.19437},
}

parthpower committed:
feat(aiob/deepseek): add AIOB for DeepSeek

reusing AiobMegatron for overall structure, and customizing MoEMLP layer
to use FP8 DeepGEMM kernels

parthpower committed:
fix(dockerfile): fix typo in dockerfile

parthpower committed:
feat(deepseek): install deep_gemm in Dockerfile

parthpower committed:
feat(deepseek): add docs, and scripts

- add deepseek examples in README
- add deepseek frame option in megatron_gpt.sh and megatron_workload_with_aiob.sh

parthpower committed:
fix(aiob/LayerNorm): fix FastLayerNorm kernel size limitations

Fused layer norm from NVIDIA APEX supports only selective sizes (listed here https://github.com/NVIDIA/apex/blob/4bdecd06b3c4b2c0a8fb6603829a8f9f05a42b49/apex/contrib/csrc/layer_norm/ln_fwd_cuda_kernel.cu#L73-L227)

this does not include hidden sizes for some model e.g. DeepSeek

In case of unsupported kernel size, fallback to pytorch's default layer_norm implementation

MXtremist commented:
After reviewing the source code, I recommend simulating DeepEP's communication performance for the DeepSeek MoE layer. The current communication method and data volume for token dispatch and combine are quite different from DeepEP's. 

Additionally, since `flashmla` is not yet supported, It seems like implementing workload generation for the Qwen-MoE model would be pretty straightforward. What do you think?

parthpower commented:
@MXtremist Thank you for trying it out

About `FastLayerNorm` questions,

1. Is this fallback expected for this configuration (e.g., for a hidden_size of 10944)?

It is expected because APEX's `FastLayerNorm` only supports [very selective sizes](https://github.com/NVIDIA/apex/blob/4bdecd06b3c4b2c0a8fb6603829a8f9f05a42b49/apex/contrib/csrc/layer_norm/ln_fwd_cuda_kernel.cu#L73-L227) (as mentioned in https://github.com/aliyun/aicb/pull/41/commits/73b6c42991892d00fa57a2f9a1df73442695ba6c)

2. Does using the standard `layer_norm` affect the accuracy of the final result?

Since FastLayerNorm doesn't support those hidden sizes, it'd be hard to say. But from my experiment (code [here](https://github.com/parthpower/aicb/commit/8bd2881188fdec698cc7b2300f4ca31406feae6e)), it seems like pytorch's  layer_norm is faster on H100

```shell
$ python -m workload_generator.mocked_model.layernorm
hidden: 4096 t_fused: 67.39 t_torch: 44.05 diff: -34.63% norm_diff: 1.05
hidden: 5120 t_fused: 53.56 t_torch: 31.84 diff: -40.55% norm_diff: 1.16
hidden: 6144 t_fused: 48.27 t_torch: 30.30 diff: -37.22% norm_diff: 1.30
hidden: 8192 t_fused: 47.25 t_torch: 31.36 diff: -33.63% norm_diff: 1.49
hidden: 12288 t_fused: 56.17 t_torch: 36.08 diff: -35.76% norm_diff: 1.87
hidden: 16384 t_fused: 104.29 t_torch: 41.16 diff: -60.53% norm_diff: 2.05
```

About larger model fitting issue, I faced similar problem and my solution was to reduce number of experts and layers such that it'd emulate EP and PP. 

In AiobMegatron, we collect times for Attention and MLP for each layer [here](https://github.com/aliyun/aicb/blob/9ac2f0ecc233996e8da3f34b2ae4c15f81543134/workload_generator/mocked_model/AiobMegatron.py#L67-L134), then process the min/max/avg based on them [here](https://github.com/aliyun/aicb/blob/9ac2f0ecc233996e8da3f34b2ae4c15f81543134/utils/utils.py#L350-L396). So possibly we can run Aiob measurements multiple times with limited layer counts and merge the results from multiple runs. It would affect timings a bit because of kernel placement layouts but it may not be significant.

parthpower commented:
> After reviewing the source code, I recommend simulating DeepEP's communication performance for the DeepSeek MoE layer. The current communication method and data volume for token dispatch and combine are quite different from DeepEP's.

Did you mean this for the AIOB or non-AIOB part (or both)? I'll run DeepEP and update my findings.

I've looked at the DeepEP's code and most of the difference compared to the one implemented on aicb, was in the group limited dispatch, and FP8 dispatch. It still seem to do an all-to-all for both dispatch/combine.

Group limited dispatch would be an overall workload property, so it'd not really change volume or comm method. I had FP8 dispatch part addressed in the MockedModel. Rest of DeepEP seemed similar to the one in aicb. Do you see any other major differences?

> Additionally, since `flashmla` is not yet supported, it seems like implementing workload generation for the Qwen-MoE model would be pretty straightforward. What do you think?

I haven't fully looked into Qwen-MoE but at first glace at the [paper](https://arxiv.org/abs/2505.09388) it seems to have GQA similar to LLAMA with addition of a QK-norm and MoE looks similar to that of LLAMA. Should be doable, possibly in a different PR.

Looking at FlashMLA's test code [here](https://github.com/deepseek-ai/FlashMLA/blob/261330bb6dfacdff8ff4b67e126417863b31aa72/tests/test_fmha_sm100.py), just adding it to AiobDeepSeek should be simple, would need to find some Blackwell GPUs to test it :)

MXtremist commented:
> Did you mean this for the AIOB or non-AIOB part (or both)? I'll run DeepEP and update my findings.
> 
> I've looked at the DeepEP's code and most of the difference compared to the one implemented on aicb, was in the group limited dispatch, and FP8 dispatch. It still seem to do an all-to-all for both dispatch/combine.
> 
> Group limited dispatch would be an overall workload property, so it'd not really change volume or comm method. I had FP8 dispatch part addressed in the MockedModel. Rest of DeepEP seemed similar to the one in aicb. Do you see any other major differences?

Thanks for the clarification. I was referring to the ​**non-AIOB part**​, specifically the MoE layer within the workload generator, i.e. `workload_generator/AIOB_simAI_workload_generator.py`. Since this section wasn't modified, it currently uses the previous MoE workload generation logic from SimAI. I noticed two things there:


1. **Communication Volume:** Since the dispatch in DeepEP uses FP8, I think the communication volume should be `tp_comm_size * self.topk // self.tp // 2`. The current calculation seems to overlook this.
2. **Communication Method:** When `tp != 1`, the existing code generates 7 communication calls. I'm not an expert on this part, but my assumption was that this should correspond to the forward and backward passes of the three main communication steps (dispatch, all-reduce, combine).

Let me know if my understanding is off here.

parthpower commented:
@MXtremist Thank you for reviewing :)

> 1. **Communication Volume:** Since the dispatch in DeepEP uses FP8, I think the communication volume should be tp_comm_size * self.topk // self.tp // 2. The current calculation seems to overlook this.

Thanks! I had updated it on the [MockedDeepSeek.py](https://github.com/aliyun/aicb/blob/407078f63e1a7fbdd455880dab6662c2047693b1/workload_generator/mocked_model/MockedDeepSeek.py#L280-L342) for FP8 dispatch but not for the SimAI part, I will update the patch for it.

It should dispatch FP8 tensor of `seq * hidden * batch // tp` along with a `/128` scaling tensor of size `seq * hidden * batch // tp // 128` (based on DeepGEMM, will add proper reference to it as well)

> 2. Communication Method: When tp != 1, the existing code generates 7 communication calls. I'm not an expert on this part, but my assumption was that this should correspond to the forward and backward passes of the three main communication steps (dispatch, all-reduce, combine).

> update 2025.09.09: I dug into the Megatron-LM code today and figured it out. Our workload generator is based on an old Megatron implementation (like v0.8.0). The newer versions now use that dispatch -> all-reduce -> combine pattern.

Thank you for verifying it :)


> The attention layer for DeepSeek is split into multiple smaller layers (at [here](https://github.com/parthpower/aicb/blob/deepseek-workload/workload_generator/mocked_model/MockedDeepSeek.py#L130-L192)). However, the workload generator seems to assign the entire Megatron attention computation time to each of these new layers (with 'row'/'column' layers getting half). For example, a layer like "attention_linear" gets its compute time from the compute_cache via the "attention_forward" key. But this key holds the total time for all AIOB operators prefixed with "atten".

Thanks for catching that bug! I haven't implemented the AIOB part for attention (MLA) block for DeepSeek, and it's reusing attention block from AiobMegatron. I can create one in pytorch to collect compute times for individual layers to fix this bug. What do you think?

parthpower committed:
feat(deepseek): fix EP dispatch/combine for FP8

For MockedDeepSeek.py and AIOB_simAI_workload_generator:

TP comms stays at BF16

EP dispatch FWD all-to-all : (batch * hidden * seq // tp) FP8 with /128 FP32 scale factor
EP dispatch BKWD all-to-all: (batch * hidden * seq // tp) BF16
EP combine (both FWD BKWD) all-to-all : (batch * hidden * seq * ep // tp) BF16

based on DeepEP test here: https://github.com/parthpower/DeepEP/commit/50aee15f592bc22142eb04b7d718296b19613ae9
(also see section 3.3.3 of https://arxiv.org/abs/2412.19437)

fixes https://github.com/aliyun/aicb/pull/41#issuecomment-3264466320

parthpower committed:
fixup! feat(deepseek): fix EP dispatch/combine for FP8

EP dispatch sizes should multiply by topK not EP ranks

parthpower committed:
feat(aiob/deepseek): add DeepSeekMLA in AIOB

- Use pytorch to create MLA AIOB layer, and use it for DeepSeek AIOB frame
- MLA publishes compute time for each layer similar to MockedDeepSeek. Partially addressing https://github.com/aliyun/aicb/pull/41#issuecomment-3274875680 for DeepSeekMLA
- Update AIOB_simAI generator to use per-layer compute times

Note:
- not adding positional embeddings stage for q_pe and k_pe because of tensor shape issue but that shouldn't affect timing as much
- AiobDeepSeek sets default torch device to "cuda:0" because of some internal tensors allocated on cpu, but this should be fine?

tianhao909 commented:
Hi @parthpower 👋

Great work on this PR! By the way, I remember you mentioned earlier about doing a tech-talk on AICB/SimAI at SCaLE (Southern California Linux Expo). Since SCaLE 23x just wrapped up earlier this month (March 5-8), I was curious — did the talk end up happening? If so, we'd love to hear how it went! And if there are any slides or recordings, it would be awesome to share them with the community.

Thanks again for your contributions! 🙏
